### PR TITLE
fix(mcdu): Fix input issues on PERF/W&B and fuel prediction pages

### LINKED ISSUE #3321: [BUG] Alternate fuel entry greater than 3.0 lbs displays "Entry out of range"
**Mod Version**
{
  "built": "2021-02-03T19:03:44+00:00",
  "ref": "master",
  "sha": "b9245c886074dc9a8a8e6cdfd7724a5a3f8077c5",
  "actor": "Benjozork",
  "event_name": "manual"
}

**Describe the bug**
When entering the actual amount of alternate fuel (ie. 4.2 for 4,200 lbs) on the INIT B page, I get an "ENTRY OUT OF RANGE" message on the scratchpad.

**To Reproduce**

1. Powered up the aircraft at gate on batteries, then GPU.
2. Aligned IRUs and verified accurate position.
3. Entered KDFW/KSTL on INIT A, KMEM as alternate, CI of 20, CRZ of FL330.
4. Within INIT B page, clicked 1R twice to enter current ZFW/CG and entered block fuel.
5. Entered desired alternate fuel of 4.2 (4,200 lbs) into the scratchpad and pressed 4L to enter that value into the ALTN blue placeholder and received the "ENTRY OUT OF RANGE" message on scratchpad.

**Expected behavior**
It should have displayed a large 4.2, followed by a "/" and a green time in minutes, calculated from that entered value.

**Actual behavior**
Instead, a value of 4.2 gave the "ENTRY OUT OF RANGE" message. Accepted values are within the range of 0.1 - 3.0. Please read below for additional context and explanation.

**References**
ALT-TIME-bug-github-screen

**Additional context**
Entering a value of 0.1 results in 0.5 lbs/3 minutes of reserve fuel. Entering a value of 3.0 results 14.6 lbs/-2856 minutes of reserve fuel. Anything outside of this range results in "ENTRY OUT OF RANGE." It appears the amount entered is being manipulated during insertion somehow.

Was this working before/when did the issue start occurring?
It has been an issue for at least a month.

Is this a problem in the vanilla unmodded game?
This feature not available in the vanilla unmodded game.

Discord username (if different from GitHub):

### LINKED ISSUE #3525: [BUG] PERF/W&B page does not handle input values correctly when imperial units are used
**Mod Version**

```json
{
  "built": "2021-02-14T01:44:40+00:00",
  "ref": "master",
  "sha": "da1f42236b04235a87c57966017050a9241ebf42",
  "actor": "Benjozork",
  "event_name": "manual"
}
```

**Describe the bug**

When entering the amount of fuel to be loaded on the PERF/W&B page in the MCDU, the value is assumed to be in kilograms even if the airplane is set up to use pounds as the weight unit.

Possibly related to #3321

**To Reproduce**

1. Set 'LBS' as the weight unit (MCDU MENU -> OPTIONS -> FMGC -> WEIGHT UNIT)
2. Reload the aircraft
3. Enter 1000 as BLOCK FUEL (MCDU MENU -> ATSU -> AOC MENU -> PERF/W&B -> Page 1)
4. Enter 1000 as PAYLOAD (MCDU MENU -> ATSU -> AOC MENU -> PERF/W&B -> Page 2)

**Expected behavior**

Expected BLOCK FUEL value to be 1000 (lbs)
Expected PAYLOAD value to be 1000 (lbs)

**Actual behavior**

BLOCK FUEL and PAYLOAD is set to 2205 (1000 kgs converted to 2205 lbs)

**References**

fuel_conversion_bug

**Additional context**

Was this working before/when did the issue start occurring?
Unknown

Is this a problem in the vanilla unmodded game?
N/A

Discord username (if different from GitHub):
Jethro#4431

### LINKED ISSUE #3528: [BUG] PERF/W&B page allows loading more fuel than max capacity indicated by MSFS and ECAM
**Mod Version**

```json
{
  "built": "2021-02-14T01:44:40+00:00",
  "ref": "master",
  "sha": "da1f42236b04235a87c57966017050a9241ebf42",
  "actor": "Benjozork",
  "event_name": "manual"
}
```

**Describe the bug**

The MCDU PERF/W&B page allows you to load up to 21273 kg of fuel, however the max fuel capacity as indicated by MSFS is only 19046 kg.

**To Reproduce**

1. Request 19046 kg of fuel to be loaded (MCDU MENU -> ATSU -> AOC MENU -> PERF/W&B -> Page 1, BLOCK FUEL and then LOAD)
2. Verify MSFS reports 19046 kg of fuel
3. Verify ECAM reports 18960 kg of fuel
4. Request 21273 kg of fuel to be loaded
5. Verify fuel levels in MSFS and ECAM unchanged
6. Request >21273 kg of fuel to be loaded
7. Verify MCDU responds NOT ALLOWED

**Expected behavior**

I am not sure about the total capacity that should be allowed bu the MCDU. There is some more in-depth discussion about usable vs unusable fuel capacity in #2286. However, I don't think the MCDU should allow you to load more than the usable (19051 kg) and unusable (65.9 kg) fuel capacity combined (19116.9 kg). 

**Actual behavior**

The MCDU allows you to load more than the usable + unusable fuel capacity (as described in #2286)

**References**

After requesting 19046 kg:

insim_fuel_19046l_with_fob_log

After requesting 21273 kg:

insim_fuel_21273l_with_fob_log

**Additional context**

Was this working before/when did the issue start occurring?
Unknown

Is this a problem in the vanilla unmodded game?
N/A

Discord username (if different from GitHub):
Jethro#4431

### PR DESCRIPTION
**Pull request by felixharnstrom**
Fixes #3525
Fixes #3321
Fixes #3528

## Summary of Changes

Fixed a number of missing or incorrect unit conversions relating to fuel and payload weight conversions when using imperial units. This fixes a number of issues where entered values were either not stored correctly or not displayed correctly. The following fields are fixed:

- PERF/W&B
   -  Page 1
      - BLOCK FUEL
      - TAXI FUEL
      - TRIP FUEL
   - Page 2
     - PAYLOAD
     - FWD BAG
     - REAR BAG
- Fuel prediction pages (INIT page B and the FUEL PRED page)
   - TAXI (on INIT B)
   - RTE RSV/%
   - ALTN/TIME
   - FINAL/TIME
   - MIN DEST FOB
   - EXTRA/TIME

Max allowed block fuel value is also changed to 19046 kg to match MSFS fuel dialog and actual fuel capacity.
Also fixes an issue where TRIP FUEL input on the PERF/W&B page was incorrectly bound to 2R.
## Screenshots (if necessary)

Before             |  After
:-------------------------:|:-------------------------:
fuel_conversion_bug | fuel_conversion_fixed
payload_conversion_bug | payload_conversion_fixed
fuel_pred_bug | fuel_pred_fixed

## References

Max fuel capacity from sources in #2286, adjusted to 19046 kg to account for rounding in MSFS

## Additional context

Discord username (if different from GitHub): Jethro#4431

## Testing instructions

Test this PR with both imperial and metric units (MCDU MENU -> OPTIONS -> FMGC -> WEIGHT UNIT, make sure to reload the aircraft after changing units): 
1. Make sure that fuel and payload values on PERF/W&B (MCDU MENU -> ATSU -> AOC MENU -> PERF/W&B) are entered correctly and that the correct amount of fuel is loaded when requesting refueling
2. Enter FROM/TO and ALTN on INIT page A 
3. Verify that both fuel prediction pages (INIT page B and FUEL PRED) behave as expected when changing field values

Note that this PR is focused around input validation and unit conversions (with some corresponding improvements to the presentation on the fuel prediction pages). Please bear this in mind when reporting potential issues with calculated values or any underlying algorithms. Also note that I have attempted to make minimal changes to the PERF/W&B payload page since this will be overhauled in an upcoming PR.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by donstim** [comment]
Thank you for proposing a fix to this issue. There are still issues with units conversion, and it is difficult to tell exactly what MSFS uses in each case. See the following screenshots for example.  First, the MSFS fuel capacities are given in US gallons in the flight_model.cfg file and are reproduced in the MSFS weight and balance window:
Screenshot (656)
Showing the same max fuel amount in lbs:
Screenshot (655)
MSFS supposedly uses 6.7 lb/gal for fuel density. However, if one does it on a tank by tank basis, the density appears to be 6.69999541 for the center tank, 6.699994493 for the inner tanks, and 6.69995614 for the outer tanks. Now we get to liters:
Screenshot (657)
For the center tank, the conversion appears to be 3.784997705 l/gal, right on 3.785 l/gal for the outers, and 3.78495614 l/gal for the inners.  And finally, in kg:
Screenshot (658)
For the center tank, the density is 0.802924761 kg/l, for the inners 0.802923084 kg/l, and for the inners 0.802924783 kg/l.

I'm not sure how to make sense of all this, but the constant for max allowable fuel of 21273 kg is incorrect. As can be seen from the last screenshot, the max allowable fuel in kg should be 19,046 kg. In pounds, it should be 41,989 lb (screenshot 2). Even when I have the units set to lbs in the MCDU options, the max fuel weight value this PR allows me to insert is 21,273.

**Comment by felixharnstrom** [comment]
I agree that the 21273 kg fuel limit feels a bit odd. I have actually already opened a separate issue about that here: #3528

Regarding your last point, are you saying that it does not allow you to enter more than 21273 even when set to use LBs? That seems odd to me and not really in line with what I've seen when testing out the PR on my end. For reference, here is what I'm seeing with this PR applied:

KG             |  LB
:-------------------------:|:-------------------------:
max_fuel_kg_fixed | max_fuel_lb_fixed

As you can see, the limit is 21273 when using KGs and 46898 when using LBs.

Did you make sure to reload the aircraft after changing it to use LBs? If not, the system is still in metric until you do. Let me know if you find the time to try it out and see if it's still an issue after reloading the aircraft.

**Comment by donstim** [comment]
Ok, I remember that I did not reload the airplane. Thanks for getting back to me.

**Comment by MisterChocker** [comment]
Whats the status of this?

**Comment by felixharnstrom** [comment]
> Whats the status of this?

I would say that this is ready for QA. I was under the impression that someone else needed to label this as ready for review first though.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_OFPData.js [hunk 1]
```diff
@@ -137,34 +138,32 @@ class CDUAocOfpData {
...
             if (value === FMCMainDisplay.clrValue) {
-                mcdu.aocWeight.tripFuel = "";
+                mcdu.aocWeight.tripFuel = undefined;
                 updateView();
                 return true;
             }
-            const maxAllowableFuel = 21273;
-            const enteredFuel = Math.round(+value);
+            const enteredFuel = Math.round(+value) / mcdu._conversionWeight;
             if (enteredFuel >= 0 && enteredFuel <= maxAllowableFuel) {
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_OFPData.js:165]: isn't `maxAllowableFuel` undefined at this stage?

Comment (felixharnstrom) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_OFPData.js:165]: My understanding was that arrow functions 'inherit' their parent's scope (see line 178 where we access `currentBlockFuel` for instance): https://medium.com/@iampika/es6-arrow-functions-syntax-and-lexical-scoping-d061732071e7

However, I'm not a Javascript expert so I might be misunderstanding something. If we're not sure if it's safe I can certainly revert this.

Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_OFPData.js:165]: I see, I am not an JS expert either, so I'll wait for @beheh response.

**Comment by donstim** [comment]
> I agree that the 21273 kg fuel limit feels a bit odd. I have actually already opened a separate issue about that here: #3528
> 
So there will be a different PR to fix #3528 ?

**Comment by felixharnstrom** [comment]
> > I agree that the 21273 kg fuel limit feels a bit odd. I have actually already opened a separate issue about that here: #3528
> 
> So there will be a different PR to fix #3528 ?

Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_OFPData.js (hunk 1 above)
Comment (beheh) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_OFPData.js:165]: Yeah, it's inherited downwards, so this code is fine.

**Comment by donstim** [comment]
> > So there will be a different PR to fix #3528 ?
> 
> Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
> That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.

Okay, but the max values are clear in the screenshots I provided above.  Those screenshots show what the max values should be.  They are correct (other than the apparent Asobo bug regarding unusable fuel).

**Comment by felixharnstrom** [comment]
> > > So there will be a different PR to fix #3528 ?
> > 
> > 
> > Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
> > That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.
> 
> Okay, but the max values are clear in the screenshots I provided above. Those screenshots show what the max values should be. They are correct (other than the apparent Asobo bug regarding unusable fuel).

Ah, gotcha. I was just a bit hesitant since I wasn't sure if 19049 (or 19051 according to your PR #2286, this might be due to rounding differences?) was what the value _should_ be. Then, unless you feel that it's better to open a separate PR, I wouldn't mind fixing that while I'm already poking around at this.

**Comment by donstim** [comment]
> > > > So there will be a different PR to fix #3528 ?
> > > 
> > > 
> > > Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
> > > That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.
> > 
> > 
> > Okay, but the max values are clear in the screenshots I provided above. Those screenshots show what the max values should be. They are correct (other than the apparent Asobo bug regarding unusable fuel).
> 
> Ah, gotcha. I was just a bit hesitant since I wasn't sure if 19049 (or 19051 according to your PR #2286, this might be due to rounding differences?) was what the value _should_ be. Then, unless you feel that it's better to open a separate PR, I wouldn't mind fixing that while I'm already poking around at this.

Yeah, it's MSFS rounding differences. Gotta go with MSFS value for it since the units conversion is in the code. The base value in the cfg file that we have control over is in gallons. That is then converted by MSFS for the other units -- lbs, liters, and kgs.

**Comment by felixharnstrom** [comment]
> Yeah, it's MSFS rounding differences. Gotta go with MSFS value for it since the units conversion is in the code. The base value in the cfg file that we have control over is in gallons. That is then converted by MSFS for the other units -- lbs, liters, and kgs.

Alright, thanks for the info :) I updated the max allowed fuel to match MSFS.

**Comment by wpine215** [comment]
Is this ready for QA testing again?

**Comment by felixharnstrom** [comment]
> Is this ready for QA testing again?

Personally I think so - there is nothing else I would like to add to the PR at least :)
But since this is my first contribution to this project I'm not sure who has the final word on that. I'm guessing only members of the flybywiresim organization?

**Review by Benjozork — APPROVED** [review]
LGTM

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 11.3.2021
Version of the sim: 1.14.5.0
PR Tested: #3545
Tier of testing: 1
Changes to observe: Fuel and payload behaviour
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power, set unit to kg
3.) Played with PERF/W+B
4.) Played with Init B

Media:


Image1

Issues: 
- On small AOC Payload FWD BAG can get negative (Image1)
- FS Fuel page baggage gets not updated
- Fuel gets burned on only bat. and ext. power, about 40kg a minute


Overall Rating: Needs to be improved

Conclusions of the testing: Lots of stuff is working. Small issues found.

Hope this helps :-)

**Comment by DavidEGTE** [comment]
Category: QA Tester
Name: David#2177
Date of testing: 11/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3545 
Tier of testing: 1

Changes to observe:
Fuel and payload behaviour

Testing technique: Try-out

Steps you've made to test the Pull Request:
1. Spawned C&D @ LIRN
2. Turned on batt + lights
3. Tested Fuel Pred 
4. Tested Perf/W+B
5.

Flight notes (When done a full flight):
N/A
Media: Insert screenshots, recordings, or anything relevant to the review here.
N/A

Issues (If Any): Write any issues
same as @marcman86 for point 1&2, point 3 I don't see fuel consumption, tried with fuel pump on and off
(tried for 45min spawn since I was watching a series in the meanwhile)
Steps to achieve the issue:
N/A

Overall Rating: To be improved
Conclusions of the testing: Most of the stuff mentioned are working, same needs to be recheck

**Comment by marcman86** [comment]
To get the "fuel leak":

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power.
3.) Turned on fuel pumps

4.) Set AOC block fuel to 100, leave taxi and trip fuel open
Here you need watch ECAM fuel page
5.) Clicked on Refuel load
ECAM shows outer tanks 40 and FOB 20kg. Just wait a little and is goes all down to 0

6.) Set AOC block fuel to 1000, leave taxi and trip fuel open
7.) Clicked on Refuel load
ECAM shows outer tanks 480 and FOB 920kg. Just wait a little and is goes down

8.) Set AOC block fuel to 4000, leave taxi and trip fuel open
9.) Clicked on Refuel load
ECAM shows outer tanks 680, inner tanks 1280 and FOB 3920kg. Just wait a little and is goes down

10.) Set AOC block fuel to 5000, leave taxi and trip fuel open
11.) Clicked on Refuel load
ECAM shows outer tanks 680, inner tanks 1780 and FOB 4920kg. Here it gets stable!

**Comment by felixharnstrom** [comment]
@marcman86  Thanks for investigating this issue :) I just verified that the same behavior exists when setting the fuel quantity via the MSFS fuel dialog as well, so it's not an input issue with the MCDU. I'm guessing it might be modelling fuel flowing from the tanks to the engines, but I might be wrong about that. Either way I think changing this is outside the scope of this PR since it relates to the underlying fuel consumption rather than being an input issue. Also, I'm guessing this might change with the upcoming changes to the engine model.

**Comment by felixharnstrom** [comment]
@marcman86 @DavidEGTE Thank you both for reviewing this. Regarding your other comments:

- The negative payload weights seem to be an existing issue: #3776. So since this is not directly caused by any of the changes made in this PR I would prefer not fixing this here since the balancing algorithm will change fairly drastically in #3474 and I would prefer not doing any more substantial changes that will just be obsolete soon (the input handling is one of the few things that might get reused) :)
- The payload not being updated in the MSFS payload menu seems to be an Asobo bug: https://github.com/flybywiresim/a32nx/pull/1865#issuecomment-745271247. You can verify that the GW value on the ECAM changes and that the suspension gets compressed after requesting the payload.

**Comment by donstim** [comment]
> To get the "fuel leak":
> 
This is a known bug in the MSFS outer to inner fuel tank transfer. If the inner tank is below the trigger value, then transfer from the outer tanks will start. However, if the engines are not on, the fuel doesn't actually make it way into the inner tanks. It just disappears.

Once the transfer starts it continues until the airplane is refueled. This is indicated in our airplane by adding enough fuel to an inner tank to give it a quantity of 474 gallons.

**Comment by St54Kevin** [comment]
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 19/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3545 
Tier of testing: 1
Changes to observe: MCDU AOC PERF/W&B page

**Steps you've made to test the Pull Request:**
1. Loaded up C&D
2. Tested different kg inputs on AOC and INIT B page
3. changed to lbs, reloaded plane
4. Tested again inputs with lbs

**Media:**
image

**Issues (If Any):** 
- I can confirm the disappearing fuel with smaller load as mentioned above

**Conclusions of the testing:** LGTM. Overall switching units doesnt seem to cause any issues anymore. Nice fix

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 19/3/21
Version of the sim: 1.14.5.0
PR Tested: #3545 
Tier of testing: 1
Changes to observe:
PERF/W&B and fuel prediction

Testing technique: Loaded in and tried
Steps you've made to test the Pull Request:
1. Loaded in
2. filled the PERF/W&B page
3. filled init a and flight plan page
4. filed init b 
5. looked at fuel pred page

Media: Insert screenshots, recordings, or anything relevant to the review here.
image
image
image
https://user-images.githubusercontent.com/70103461/111842857-090c4c80-88f8-11eb-9319-82884e863b0d.mp4
https://user-images.githubusercontent.com/70103461/111840843-f0e6fe00-88f4-11eb-9bc4-3306600a1130.mp4
https://user-images.githubusercontent.com/70103461/111840881-fd6b5680-88f4-11eb-92f3-43d6743a0184.mp4
https://user-images.githubusercontent.com/70103461/111840911-09571880-88f5-11eb-96df-9fb705687dbd.mp4
https://user-images.githubusercontent.com/70103461/111843103-7ddf8680-88f8-11eb-9cea-4f0acc6c1b93.mp4

Issues (If Any): 
- more of a suggestion but maybe make the min you are able to put into the block fuel area the 1% in the msfs menu (currently lets you enter even 1 as shown in the screenshot
- FWD baggage goes into minus when the payload is from 1 to 701
- negative values can be entered in both FBW and rear baggage
- PAX number cannot be entered
- When entering RTV RSV it doesn't enter and shows no message but % is allowed
- When entering a negative into MIN DEST it says FORMAT ERROR but when entered into FINAL or ALTN it says ENTRY OUT OF RANGE. Then when entering a negative into RTE RSV/% it doesn't say anything or enter

Steps to achieve the issue
do as shown in the Media section above

Overall Rating: To be improved
Conclusions of the testing: Quite a few things I have found which may be realistic or how they are supposed to be but just pointing everything that seems wrong to me.

**Comment by felixharnstrom** [comment]
@FlightSimmer7700 Thank you for the in-depth review.

- I would prefer not adding any logic to the minimum fuel amount based on the MSFS fuel dialog. I simply don't see any reason why we should limit ourselves to such arbitrary restrictions in the MSFS UI unless we really need to.
- I added a check to make sure that the entered payload values are positive. Other than that I would prefer not to make any more changes to the payload page since it will change fairly drastically in an upcoming PR as I described in a previous comment. That includes the inoperable PAX field and the negative calculated payload values.
- I fixed the error message when entering a negative value in MIN DEST. It should now say ENTRY OUT OF RANGE like the other fields.

I haven't done any changes to the RTE RSV field yet. I'll need to dig around a bit and try to determine how it should behave before I make any changes to the input logic.

**Comment by FlightSimmer7700** [comment]
> @FlightSimmer7700 Thank you for the in-depth review.
> 
> * I would prefer not adding any logic to the minimum fuel amount based on the MSFS fuel dialog. I simply don't see any reason why we should limit ourselves to such arbitrary restrictions in the MSFS UI unless we really need to.
> * I added a check to make sure that the entered payload values are positive. Other than that I would prefer not to make any more changes to the payload page since it will change fairly drastically in an upcoming PR as I described in a previous comment. That includes the inoperable PAX field and the negative calculated payload values.
> * I fixed the error message when entering a negative value in MIN DEST. It should now say ENTRY OUT OF RANGE like the other fields.
> 
> I haven't done any changes to the RTE RSV field yet. I'll need to dig around a bit and try to determine how it should behave before I make any changes to the input logic.

Sounds good, I'll take a look at the pr later today and check that everything is working. Apart from the RTE RSV

**Comment by FlightSimmer7700** [comment]
Ok, I tried it out again. Cannot input negative values but they still come up when putting low payload(I think that will be fixed in the future as you said). I found the tiniest issue which I'm not sure is even worth commenting and then working on but when entering a negative into min dest you get the ENTRY OUT OF RANGE message but then when you clear that the number is still there while when doing it on the others above the ENTRY OUT OF RANGE comes up and then when cleared the number does not show again

https://user-images.githubusercontent.com/70103461/111888320-514c6d00-89d3-11eb-8acb-a735b134679f.mp4

**Comment by felixharnstrom** [comment]
@FlightSimmer7700 Good catch! I pushed a fix for that issue - now the value is cleared like for the other fields.

I've also made a number of changes to the RTE RSV/% and FINAL/TIME fields. Both of these fields now:
- Disallow entry in the `XX/YY` format. The Pilot's guide how only one or the input types shuld be allowed at a time, not both at the same time. So now `XX/` (the trailing slash is optional) is treated as weight input, and `/YY` is interpreted as percentage for the route resereve and time for the final fuel.
- Made the input parser a lot stricter overall. For instance, the previous implementation allowed input like `/0.2/ABC`. Now the only allowed input formats are `XX/`, `XX` and `/XX`.
- The manually entered value is more consistently displayed in large font and the calculated value in small font.

I did not change how the RTE RSV/% field is still replaced with all slashes when an out of range value is entered.

**Review by Lucky38i — APPROVED** [review]
LGTM, just needs QA to test all possible entries and void values

**Comment by felixharnstrom** [comment]
@MisterChocker This PR will actually not fix the issue that you linked (#3776) 🙂  That is an issue with the underlying payload distribution algorithm: If you enter a small enough payload value, the calculated rear and forward baggage values can be negative. I have not fixed this in this PR since this is not really an issue with the input handling per se and that page will change fairly drastically in #3474 anyway. All I've done in this PR is make sure that the user cannot enter any negative values directly.

**Comment by TheSquirtingPantin** [comment]
Hello All, 

Category: QA Trial
Name: Discord User Rémi#8871
Date of testing: Current date 23/03/2021
Version of the sim: V 1.14.5.0
PR Tested: #3545
Tier of testing: 1
Changes to observe:
Logic of PERF/WB and FUEL PRED pages

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. C/D at KDEN
2. Check for void and logic in PERF/WB Page
3. 
4.
5.

Flight notes (When done a full flight):N/A

Media: N/A

Issues (If Any):
	- Entered random value for "Payload" in PERF/WB 2/2 page (12000kg). Then selected OFP SEND REQUEST (from SIMBRIEF with 160 pax). Clicking on 1L is calculating new Payload and calculation for payload done with negative rear bags
	- Separation bags and passengers confusing (why not renaming "passenger load", "bag fwd" and "rear bag". Payload being the sum of the 3? Is it a limitation from MSFS?) 
	- When importing the data from OFP, number of passenger is green (imported data). If entering a random number in 3R, then clicking on the 1L "Payload", the data previously registered in 3R is shown (despite 3R being still locked green).

Overall Rating: To be improved
Conclusions of the testing: NA

**Comment by felixharnstrom** [comment]
@TheSquirtingPantin Thank you for the review. As I mentioned in a previous comment, the PERF W/B payload page will be overhauled in an upcoming PR. As such I would prefer not making any more changes to this page than the basic input validation and unit conversions already included in this PR since any such changes will soon be obsolete.

**Comment by BritishAvgeek** [comment]
Category: QA Trial
Name: BritishAvgeek#1436
Date of testing: 24/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3545
Tier of testing: 1
Changes to observe: Input changes to the Perf/W&B and Fuel Prediction pages 

Testing technique: 
Spawned C&D and input all data for a routine flight and edited the values multiple times.
Reloaded to lbs to check same inputs as a comparison to see if any issues arise 

Steps you've made to test the Pull Request:
1. Noted all comments above in relation to focus on basic input and validation so some issues as mentioned above were out of the scope of this PR
2. Spawned C&D
3. Imported flight plan from EDDB to EFHK using Simbrief 
4. Configured MCDU appropriately as "routine" to check for any problems 
5. Amended the figures input 
6. Tried to input various values above the max and the PR worked and didn't allow it in
7. Repeated the steps having reloaded with the change to lbs and followed all the above in repeat
8. Changed back to KG and reloaded, worked through the flow again to ensure no issues in the reverse of the data change back to kg


Flight notes (When done a full flight):N/A
Media: N/A

Issues (If Any):
None

Overall Rating: Good
Conclusions of the testing: The PR addresses everything it says it does and I found no issues during my tests noting everything commented on and discussed further up. Thanks!

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 28.3.2021
Version of the sim: 1.14.6.0
PR Tested: #2545
Tier of testing: 1
Changes to observe: PERF/W&B and fuel prediction page inputs
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power
3.) Tested inputs on PERF/W&B page 1&2
4.) Tested inputs on INIT B page

Media:


Image1

Issues: 

- FWD Bag still gets negative when PAYLOAD < 702
- RTE RSV/% only accepts /YY and /YY.Y
- FINAL /TIME is limited to 0130, but XX/ entry can be higher than 1.9 (Image1)

Overall Rating: Needs to be improved

Conclusions of the testing: Small bugs found

Hope this helps :-)

**Comment by felixharnstrom** [comment]
@marcman86 Thank you for the review :)

- The payload page will be overhauled significantly in an upcoming PR and since this is an issue with the underlying algorithm rather than a straight-up input issue, I would prefer not adressing this in this PR.
- I have pushed a fix for the route reserved fuel issue, thanks for spotting this!
- The final fuel weight limitation is intended. It should be between 0-100 even though this doesn't exactly correspond to the 1H30M max time:

image

**Comment by marcman86** [comment]
Retested after fix:

RTE RSV/% still not working.

It just goes ---.-/--.- 
No Entry out of range or format error message is shown

**Comment by felixharnstrom** [comment]
@marcman86 Thanks you for taking another look at it. However, I am unable to reproduce this issue. Could you maybe record a short video when you have time? Also, please make sure that you are trying out the artifact built from the latest commit.

For reference, I downloaded the latest package built from this PR from Github and tried it out. As far as I can tell, the route reserved field behaves as expected:

rte_rsv_fixed2

Contents of `build_info.json`:
```
{
  "built": "2021-03-28T11:21:06+00:00",
  "ref": "merge",
  "sha": "925d3ab4c603c535a94d74a087fade4baae6fe5b",
  "actor": "felixharnstrom",
  "event_name": "manual"
}
```

**Comment by marcman86** [comment]
I waited till the build was finished. But will try again.

**Comment by MisterChocker** [comment]
The build is finished

**Comment by felixharnstrom** [comment]
@marcman86 Thank you for taking another look and for being patient with me. Please let me know the steps to reproduce if you find that it's still an issue.

**Comment by marcman86** [comment]
Still not working. Same as before.

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power
3.) Init A EDDP/EDDB, CI10, FL160
4.) Init B Autofill ZFW/ZFWCG and did (auto-) Fuel Planing
5.) Enter "2" end Press 3L (RTE RSV/%)

shows ---.-/--.-

**Comment by felixharnstrom** [comment]
@marcman86 Thanks for taking another look, now I know what you're referring to :) If 2 tonnes (in your example) correspond to a percentage greater than 15%, replacing the field with dashes is the intended (and pre-existing) behavior.

Does the issue with the route reserved weight being 0.0 all the time seem fixed to you?

**Comment by marcman86** [comment]
Oh, ok. but with no "Out of range" message or something? Will check the other thing.

**Comment by felixharnstrom** [comment]
@marcman86 The reference is not explicit on this, but the way I read it is that the field gets dashed and no message is shown in the scratchpad. Maybe @Lucky38i can weigh in on this?

**Comment by MisterChocker** [comment]
Alright waiting for luckys response before merging.

**Comment by felixharnstrom** [comment]
Actually, I just now saw that IRL A320 pilot "polonski" have posted a super helpful reference video on Discord. It seems that percentage entry outside 0-15 *should* be disallowed and give an out of range message, and if the corresponding percentage when entering the fuel weight manually is out of range it should be allowed and just result in the percentage part being dashed, not the whole field. I'll push a new commit tomorrow that fixes this.

**Comment by felixharnstrom** [comment]
I decided that sleep is for the weak and did the changes immediately :)

My latest commit fixes a number of input and display issues on the Init B and the fuel prediction pages:
- Only the route reserved percentage field is now white and dashed when the weight is out of range instead of the whole thing.
- It is now explicitly disallowed to enter route reserved percentages outside the 0-15% range.
- The alternate fuel time is now dashed and white if no alternate destination has been set.
- The alternate fuel weight is now green and pilot entry is disallowed if no alternate destination has been set.
- Extra time is now white and dashed if the extra fuel weight is negative.

References:

*Screenshot showing the INIT B page from this Youtube video:* https://www.youtube.com/watch?v=DP8soiH_4jE&t=122s
cockpit_view

Unfortunately, the reference video recorded by Discord user and IRL A320 pilot *polonski#4524* is too large to attach to this comment. Here is the link to the post in Discord: https://discord.com/channels/738864299392630914/739223532269076561/825622509197852672
Here is a screenshot showing the most important part:
image

*Pegasus Step 1A (Rev 0), 2009 edition manual*
pilots_guide
image
image
image
image

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 29.3.2021
Version of the sim: 1.14.6.0
PR Tested: #3545
Tier of testing: 1
Changes to observe: Init B fixes
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Start up and Init A EDDP/EDDM
3.) Test Init B

Issues: -

Overall Rating: Good

Conclusions of the testing: Works. Nice work!

Hope this helps :-)

**Comment by DavidEGTE** [comment]
Category: QA Tester
Name: David#2177
Date of testing: 01/04/2021
Version of the sim: 1.14.6.0
PR Tested: #3545 
Tier of testing: 1

Changes to observe:
Fix input issues on PERF/W&B and fuel prediction pages

Testing technique: Try-out

Steps you've made to test the Pull Request:
1. Spawned C&D
2. Played with PERF/W&B random + max capacity fuel
3. Tested Init Page with some values 

Flight notes (When done a full flight):
N/A
Media: Insert screenshots, recordings, or anything relevant to the review here.
N/A
Issues (If Any): Write any issues
N/A
Steps to achieve the issue:
N/A

Overall Rating: Good
Conclusions of the testing: Works like a charm! GTG

**Comment by Benjozork** [comment]
@felixharnstrom this needs to be rebase for hard-fork compatibility.

**Comment by felixharnstrom** [comment]
@Benjozork Done!

**Comment by felixharnstrom** [comment]
Added a new section in the changelog for the 0.7.0 milestone in my latest commit. Once all the commits for 0.6.0 are in, this should be ready to go :)

**Pull request merged**